File: 연습/연산자끼워넣기/연산자 끼워넣기.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int minAns = 1000000001;
int maxAns = -1000000001;
int n;
int a[12];
void cal(int cnt, int add, int sub, int mul, int div, int res){
  if(cnt==n){
    minAns= min(minAns,res);
    maxAns=max(maxAns,res);
    return;
  }
  if(add > 0) cal(cnt+1,add-1,sub,mul,div,res+a[cnt]);
  if(sub > 0) cal(cnt+1,add,sub-1,mul,div,res-a[cnt]);
  if(mul > 0) cal(cnt+1,add,sub,mul-1,div,res*a[cnt]);
  if(div > 0) cal(cnt+1,add,sub,mul,div-1,res/a[cnt]);
}

int main(){
  cin>>n;
  for(int i=0; i<n; i++) cin>>a[i];
  int add,sub,mul,div; 
  cin>>add>>sub>>mul>>div;
  cal(1,add,sub,mul,div,a[0]);
  cout<<maxAns<<'\n';
  cout<<minAns<<'\n';

  return 0;
}

File: 연습/재귀/연산자끼워넣기/연산자끼워넣기.py
def cal(a,res,idx,add,sub,mul,div):
  if idx==len(a): return (res,res)
  restuls =[] 
  if add > 0:
    restuls.append(cal(a, res+a[idx], idx+1, add-1, sub, mul, div))
  if sub > 0:
    restuls.append(cal(a, res-a[idx], idx+1, add, sub-1, mul, div))
  if mul > 0:
    restuls.append(cal(a, res*a[idx], idx+1, add, sub, mul-1, div))
  if div > 0 : 
    if res >= 0:
      restuls.append(cal(a, res//a[idx], idx+1, add, sub, mul, div-1))
    else:
      restuls.append(cal(a, -(-res//a[idx]), idx+1, add, sub, mul, div-1))
  ans = (max([p[0] for p in restuls]), min([p[1] for p in restuls]))
  return ans
n=int(input())
a=list(map(int,input().split()))
add,sub,mul,div = map(int,input().split())
ans=cal(a,a[0],1,add,sub,mul,div)
print(ans[0])
print(ans[1])
